[WebXR] Refactor WebXROpaqueFramebuffer attachment logic

mwyrzykowski opened the pull request:
#### 42e826978f5dd46198a2fc437b42fdbe1e14fd56
<pre>
[WebXR] Refactor WebXROpaqueFramebuffer attachment logic
<a href="https://bugs.webkit.org/show_bug.cgi?id=269683">https://bugs.webkit.org/show_bug.cgi?id=269683</a>
<a href="https://rdar.apple.com/123201364">rdar://123201364</a>

Reviewed by Tim Horton.

Patches are from Dan and I added some small bug fixes.

Refactor WebXROpaqueFramebuffer handling of framebuffer attachments to support
Shared or Layered layout.

There are a few combinations of context attributes that WebXROpaqueFramebuffer
handles:

if `antialias`:
   Create internal MSAA color buffer
   if `depth/stencil`
      Create internal MSAA depth/stencil buffer
if `!antialias`:
   Draw to externally provided color texture
   if `depth/stencil`
       if `!externally provided depth/stencil`
           Create internal depth/stencil buffer

Prior to this change, the code handled these combinations in `StartFrame(...)` &amp;
`SetupFramebuffer(...)` with a loose collection of state.

This change simplifies the old assumption, dropping support for external
Textures, standardizing on Renderbuffers, and introduces helpers for keeping
related state together via `WebXRExternalRenderbuffer` and
`WebXRAttachmentSet`.

WebXRExternalRenderbuffer represents an externally provided Renderbuffer through
the OpenGL renderbuffer object and EGLImage pointer.

WebXRAttachmentSet represents a group of related color, depth and stencil
buffers. These can be &quot;internal&quot; or &quot;external&quot; via `WebXRAttachments` &amp;
`WebXRExternalAttachments`.

Object naming has been changed to consistently use `draw` for the target of
WebGL rendering commands and `display` for the target presented to the user via
a WebXR compositor system. To simplify the handling of attachments, the output
of WebGL commands are always &quot;blitted&quot; from the draw target to display target,
removing the optimal path in the previous code. This path will not be taken on
any compositors supported on Cocoa and will be redressed in a future patch.

Due to limitations of `glBlitFramebuffer`, blitting to `Layered` layout from an
MSAA enabled attachment set requires the use of an intermediate, resolved path
represented by the `resolve` prefixed `m_resolvedAttachments` &amp; `m_resolveFBO`.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::ensure):
(WebCore::createAndBindCompositorBuffer):
(WebCore::WebXROpaqueFramebuffer::WebXROpaqueFramebuffer):
(WebCore::WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer):
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
(WebCore::WebXROpaqueFramebuffer::resolveMSAAFramebuffer):
(WebCore::WebXROpaqueFramebuffer::blitShared):
(WebCore::WebXROpaqueFramebuffer::blitSharedToLayered):
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer):
(WebCore::WebXROpaqueFramebuffer::allocateRenderbufferStorage):
(WebCore::WebXROpaqueFramebuffer::allocateAttachments):
(WebCore::WebXROpaqueFramebuffer::bindAttachments):
(WebCore::WebXRExternalRenderbuffer::destroyImage):
(WebCore::WebXRExternalRenderbuffer::release):
(WebCore::WebXRExternalRenderbuffer::leakObject):
(WebCore::createAndBindCompositorTexture): Deleted.
(WebCore::WebXROpaqueFramebuffer::allocateColorStorage): Deleted.
(WebCore::WebXROpaqueFramebuffer::allocateDepthStencilStorage): Deleted.
(WebCore::WebXROpaqueFramebuffer::bindColorBuffer): Deleted.
(WebCore::WebXROpaqueFramebuffer::bindDepthStencilBuffer): Deleted.
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
(WebCore::WebXRAttachmentSet::release):
(WebCore::WebXRAttachmentSet::leakObject):
(WebCore::WebXROpaqueFramebuffer::framebuffer const):
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
(WebCore::destroyFunc):
* Source/WebCore/platform/xr/PlatformXR.h:
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
(WebKit::ARKitCoordinator::renderLoop):

Metal: Override internal format of images

Add support for overriding the internal format when binding Metal
textures to EGL images.

Fixed: angleproject:8552
Change-Id: I364a88a2e608e462c5216c92927679814ab91ca5

The following changes make blitting work.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::createAndBindCompositorBuffer):
(WebCore::WebXROpaqueFramebuffer::startFrame):
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::createAndBindEGLImage):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::createAndBindEGLImage):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::createAndBindEGLImage):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::createAndBindEGLImage):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/RemoteGraphicsContextGLProxyCocoa.mm:

Helper utilites for creating, enabled and disabling the use of rasterization
rate maps in GPUP. Infrastructure inplace to support caching of map for Shared
and Layered layouts, but only Shared is currently supported.

* Source/WebCore/PAL/pal/spi/cocoa/MetalSPI.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/GraphicsTypesGL.h:
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm:
(WebCore::newRasterizationRateMap):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::createFoveation):
(WebCore::GraphicsContextGLCocoa::enableFoveation):
(WebCore::GraphicsContextGLCocoa::disableFoveation):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(createFoveation):
(enableFoveation):
(disableFoveation):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::createFoveation):
(WebKit::RemoteGraphicsContextGLProxy::enableFoveation):
(WebKit::RemoteGraphicsContextGLProxy::disableFoveation):
* Tools/Scripts/generate-gpup-webgl:

Refactor setup logic to be passed once. WebXROpaqueFramebuffer::setupFramebuffer
should extract the necessary data to create framebuffers from the information
provided in PlatformXR::FrameData::LayerSetupData.

Changed WebXRWebGLLayer to query WebXROpaqueFramebuffer for the viewport, since
WebXRWebGLLayer has no knowledge of the structure of the layout, dimensions,
etc.

Canonical link: <a href="https://commits.webkit.org/276825@main">https://commits.webkit.org/276825@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fd5772fb1173d6dfc8a272448a8a976cc915dc43

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45794 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24922 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48377 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48464 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41831 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48100 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29209 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22319 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46372 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22002 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39508 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18666 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/45659 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19402 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/40603 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3838 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42102 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40952 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50223 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20786 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17294 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/44616 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22090 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43484 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10168 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22449 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21780 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.mm:
```diff
@@ -1548,15 +1553,42 @@ inline void CheckPrimitiveType(MTLPrimitiveType primitiveType)
 #endif  // ANGLE_MTL_SIMULATE_DISCARD_FRAMEBUFFER
 }
 
+static bool shouldUseRasterizationRateMap(auto& cachedRenderPassDescObjC)
```
@kkinnunen-apple I don't know how to better handle this so the rasterization map only gets set on the main render target, as I have no idea how to identify the main render target in an WebGL / WebXR application

We can't apply the rasterization rate map to offscreen passes and even this logic is broken if the user performs an offscreen pass which matches the double-width shared layout frame buffer size.

mwyrzykowski replied:
I will mirror what the QCOM_framebuffer_foveated extension does to associate the rasterization rate map with an FBO

mwyrzykowski commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_command_buffer.mm:
```diff
@@ -1768,12 +1800,22 @@ inline void CheckPrimitiveType(MTLPrimitiveType primitiveType)
 {
+    auto maxScissorRect =
+        MTLCoordinate2DMake(mRenderPassMaxScissorRect.width, mRenderPassMaxScissorRect.height);
+    if (!shouldUseRasterizationRateMap(mCachedRenderPassDescObjC))
```
This in an addition from Dan's changes @kkinnunen-apple, since the scissor modification should only be applied on the double-width shared layout render target

mwyrzykowski commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_utils.mm:
```diff
@@ -110,7 +110,7 @@ bool FrameCaptureDeviceScope()
 void StartFrameCapture(id<MTLDevice> metalDevice, id<MTLCommandQueue> metalCmdQueue)
 {
-#if ANGLE_METAL_FRAME_CAPTURE_ENABLED
+#if 1 // ANGLE_METAL_FRAME_CAPTURE_ENABLED
```
this should be reverted

mwyrzykowski commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_utils.mm:
```diff
@@ -124,7 +124,7 @@ void StartFrameCapture(id<MTLDevice> metalDevice, id<MTLCommandQueue> metalCmdQu
     if (captureManager.isCapturing)
     {
-        return;
+        StopFrameCapture();
```
I will revert this

mwyrzykowski commented on Source/ThirdParty/ANGLE/src/libANGLE/validationEGL.cpp:
```diff
@@ -3483,15 +3483,16 @@ bool ValidateCreateImage(const ValidationContext *val,
 
             case EGL_TEXTURE_INTERNAL_FORMAT_ANGLE:
-                if (!displayExtensions.imageD3D11Texture && !displayExtensions.vulkanImageANGLE)
+                if (!displayExtensions.imageD3D11Texture && !displayExtensions.vulkanImageANGLE &&
```
this change can be reverted, it was unintentionally added

mwyrzykowski commented on Source/ThirdParty/ANGLE/src/libANGLE/validationEGL.cpp:
```diff
@@ -3483,15 +3483,16 @@ bool ValidateCreateImage(const ValidationContext *val,
-                }
-                break;
+                    val->setError(EGL_BAD_PARAMETER,
+                                  "EGL_TEXTURE_INTERNAL_FORMAT_ANGLE cannot be used without "
```
as was this one

mwyrzykowski commented on Source/ThirdParty/ANGLE/scripts/gl_angle_ext.xml:
```diff
@@ -1376,6 +1385,7 @@
                 <command name="glLogicOpANGLE"/>
             </require>
         </extension>
+<<<<<<< HEAD
```
Will fix this, bad merge

mwyrzykowski commented on Source/ThirdParty/ANGLE/src/common/gl_enum_utils_autogen.cpp:
```diff
@@ -2786,6 +2786,7 @@ const char *GLenumToString(GLESEnum enumGroup, unsigned int value)
                     return "GL_SHADING_RATE_4X2_PIXELS_QCOM";
                 case 0x96AE:
                     return "GL_SHADING_RATE_4X4_PIXELS_QCOM";
+<<<<<<< HEAD
```
will remove this marker

mwyrzykowski commented on Source/ThirdParty/ANGLE/src/libANGLE/Context.cpp:
```diff
@@ -2903,6 +2903,16 @@ void Context::finish()
     ANGLE_CONTEXT_TRY(mImplementation->finish(this));
 }
 
+void Context::StopFrameCapture()
```
going to remove the frame capture as I found another way to take frame captures

mwyrzykowski commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/IOSurfaceSurfaceMtl.mm:
```diff
@@ -22,7 +22,7 @@
 // ANGLE_METAL_FRAME_CAPTURE flag.
 #if defined(NDEBUG) && !defined(ANGLE_METAL_FRAME_CAPTURE)
-#    define ANGLE_METAL_FRAME_CAPTURE_ENABLED 0
+#    define ANGLE_METAL_FRAME_CAPTURE_ENABLED 1
```
will revert this

mwyrzykowski commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/SurfaceMtl.mm:
```diff
@@ -686,8 +686,7 @@
 
-        ANGLE_MTL_LOG("Current metal drawable size=%d,%d", mColorTexture->width(),
-                      mColorTexture->height());
+        ANGLE_MTL_LOG("Current metal drawable size=%d,%d", mColorTexture->width(mtl::MipmapNativeLevel(0)),  mColorTexture->height(mtl::MipmapNativeLevel(0)));
```
this was added since this code did not compile when `ANGLE_MTL_LOG` was enabled

mwyrzykowski commented on Source/WebCore/Modules/webxr/WebXRWebGLLayer.h:
```diff
@@ -110,7 +110,6 @@ class WebXRWebGLLayer : public WebXRLayer, private CanvasObserver {
     bool m_antialias { false };
     bool m_ignoreDepthValues { false };
     bool m_isCompositionEnabled { true };
-    bool m_viewportsDirty { true };
```
@djg do we need this boolean? It wasn't getting updated when the viewports changed from the compositor and all it prevents is assignment of two MTLViewport instances, so its not preventing much work in the first place.

djg replied:
@mwyrzykowski This is to do with the XR spec (which I'm no expert on), I think it's to do with the dynamic scaling feature in the spec.

mwyrzykowski replied:
So the problem is that the underlying frameworks can change the viewport size with no signal from the JavaScript layer. How do we tell WebXRWebGLLayer that it needs to update the viewport in that case? Is it wrong to unconditionally update the viewports?

mwyrzykowski commented on Source/WebCore/Modules/webxr/WebXRWebGLLayer.cpp:
```diff
@@ -227,20 +229,12 @@ ExceptionOr<RefPtr<WebXRViewport>> WebXRWebGLLayer::getViewport(WebXRView& view)
-        viewportData.currentScale = view.requestedViewportScale();
-        m_viewportsDirty = true;
-    }
+    WTFLogAlways("GL DEBUG MESSAGE: WebXRWebGLLayer::getViewport %s, viewport(xy,wh) = %d,%d, %d, %d", view.eye() == XREye::Right ? "right" : "left", (int)viewportData.viewport->x(), (int)viewportData.viewport->y(), (int)viewportData.viewport->width(), (int)viewportData.viewport->height());
```
this should be removed

mwyrzykowski commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/ImageMtl.mm:
```diff
@@ -20,6 +20,30 @@
 
+namespace
+{
+angle::FormatID intendedFormatForMTLTexture(id<MTLTexture> texture,
```
@djg do you know if this was a change you added or a bad merge on my part?

hortont424 replied:
did you sort this out?

mwyrzykowski replied:
I did :) it is needed, so correct to keep the change

mwyrzykowski commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/mtl_render_utils.mm:
```diff
@@ -688,8 +688,10 @@ void SetupCommonBlitWithDrawStates(const gl::Context *context,
         GetViewport(params.dstRect, params.dstTextureSize.height, params.dstFlipY);
     MTLScissorRect scissorRectMtl =
         GetScissorRect(params.dstScissorRect, params.dstTextureSize.height, params.dstFlipY);
+    // TODO(djg): Should the rasterization rate map affect blit operations?
```
oh I will remove this comment, they should not affect blit operations

mwyrzykowski replied:
Will resolve in next update

achan00 commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -212,117 +202,295 @@ void WebXROpaqueFramebuffer::endFrame()
         return;
 
+    gl->disableFoveation();
+
```
Nit: May want to move the `disableFoveation()` call after the `WebXRLayerEndFrameStart` tracepoint so that call is included in the signpost interval.

hortont424 replied:
Should you do this?

achan00 commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -714,7 +759,11 @@ static bool needsEAGLOnMac()
+        && enableExtension("GL_NV_framebuffer_blit"_s)
+#if !PLATFORM(IOS_FAMILY_SIMULATOR)
+        && enableExtension("GL_ANGLE_variable_rasterization_rate_metal"_s);
+#endif
```
@mwyrzykowski  I made a mistake here. Should be:
```
         && enableExtension("GL_EXT_sRGB"_s)
         && enableExtension("GL_OES_EGL_image"_s)
         && enableExtension("GL_OES_rgb8_rgba8"_s)
 #if !PLATFORM(IOS_FAMILY_SIMULATOR)
         && enableExtension("GL_ANGLE_variable_rasterization_rate_metal"_s)
 #endif
         && enableExtension("GL_NV_framebuffer_blit"_s);
 }
 #endif
```

mwyrzykowski replied:
Oh yes otherwise it won’t compile for sim

djg commented on Source/WebKit/Shared/XR/PlatformXR.serialization.in:
```diff
@@ -95,13 +100,26 @@ enum class PlatformXR::XRTargetRayMode : uint8_t {
 #if PLATFORM(COCOA)
-    WebCore::IntSize framebufferSize;
+    std::optional<PlatformXR::FrameData::LayerSetupData> layerSetup;
+    uint64_t renderingFrameIndex;
```
Why was this reordered?

mwyrzykowski replied:
because its used as a HashKey and hash keys in WTF have a limit of 150 bytes which we exceeded. So its reordered in size largest -> smallest to reduce struct size due to struct padding

djg commented on Source/WebCore/Modules/webxr/WebXRWebGLLayer.cpp:
```diff
@@ -228,19 +230,10 @@ ExceptionOr<RefPtr<WebXRViewport>> WebXRWebGLLayer::getViewport(WebXRView& view)
-        viewportData.currentScale = view.requestedViewportScale();
-        m_viewportsDirty = true;
-    }
-
```
I don't think it's okay to just delete all of this code.

mwyrzykowski replied:
Why do we need the `m_viewportsDirty` boolean? It was not getting updated correctly, leading to viewports being stale and out of sync with `m_leftViewportData` and `m_rightViewportData`. Is it incorrect to always update `m_leftViewportData` and `m_rightViewportData`?

djg commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
```diff
@@ -54,38 +84,53 @@ class WebXROpaqueFramebuffer {
-    PlatformGLObject allocateDepthStencilStorage(GraphicsContextGL&, GCGLsizei, IntSize);
-    void bindColorBuffer(GraphicsContextGL&, PlatformGLObject);
-    void bindDepthStencilBuffer(GraphicsContextGL&, PlatformGLObject);
+#if PLATFORM(COCOA)
+    bool setupFramebuffer(GraphicsContextGL&, const PlatformXR::FrameData::LayerSetupData&);
+#endif
```
This code is really need for the non-Cocoa platforms too. With the countless refactoring, OpenXR is probably very broken. We should split the code to have WebXROpaqueFramebufferCocoa...

mwyrzykowski replied:
`LayerSetupData` only exists for `PLATFORM(COCOA)`. Should it be added for all platforms?

mwyrzykowski replied:
The frame buffer setup should occur in `WebXROpaqueFramebuffer::startFrame` for OpenXR

djg commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
```diff
@@ -54,38 +84,53 @@ class WebXROpaqueFramebuffer {
+    Ref<WebGLFramebuffer> m_drawFramebuffer;
     WebGLRenderingContextBase& m_context;
     Attributes m_attributes;
+    PlatformXR::Layout m_displayLayout = PlatformXR::Layout::Layered;
```
I don't think should default to `Layered`. I'd prefer to have the init code detect and flip it.

mwyrzykowski replied:
Oh we do detect already, so I can probably just revert this to Shared

djg commented on Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm:
```diff
@@ -80,6 +121,77 @@ void destroyPbufferAndDetachIOSurface(EGLDisplay display, void* handle)
+    if (!EGL_QueryDeviceAttribEXT(device, EGL_METAL_DEVICE_ANGLE, reinterpret_cast<EGLAttrib*>(&mtlDevice)))
+        return nullptr;
+
+#if PLATFORM(COCOA) && !PLATFORM(WATCHOS) && !PLATFORM(APPLETV)
```
This file is compiled under Cocoa only, so `PLATFORM(COCOA)` is redundant

djg commented on Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm:
```diff
@@ -80,6 +121,77 @@ void destroyPbufferAndDetachIOSurface(EGLDisplay display, void* handle)
+    UNUSED_PARAM(physicalSizeLeft);
+    UNUSED_PARAM(physicalSizeRight);
+    RetainPtr<MTLRasterizationRateMap> rasterizationRateMap;
+#endif
```
This is a mess. I think we should have a separate function for obtaining the rate map from CompositorServices (in a separate file).

mwyrzykowski replied:
We could remove lines 167-174 if you think that would help, they are not used other than for assertions.

mwyrzykowski replied:
I will remove these assertions, they are not needed

hortont424 commented on Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/ContextMtl.mm:
```diff
@@ -1648,6 +1653,33 @@ GLint GetOwnershipIdentity(const egl::AttributeMap &attribs)
+{
+    id<MTLRasterizationRateMap> rateMap = (__bridge id<MTLRasterizationRateMap>)(map);
+    if (rateMap && rateMap.device != mContextDevice.get())
+    {
```
IDK about angle's style rules but you've got some inconsistency here with the braces

mwyrzykowski replied:
oh thank you, they should all be on separate lines

hortont424 commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
```diff
@@ -42,6 +42,36 @@ class WebGLFramebuffer;
 struct XRWebGLLayerInit;
 
+struct WebXRExternalRenderbuffer {
+    GCGLOwnedRenderbuffer rbo;
```
Do we normally use this abbreviation? We tend to stay away from all but the most obvious abbreviations in WebKit

mwyrzykowski replied:
Oh good observation, I will rename `renderBufferObject`

hortont424 commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
```diff
@@ -42,6 +42,36 @@ class WebGLFramebuffer;
+        depthStencilBuffer.release(gl);
+    }
+
+    void leakObject()
```
Ooh what is this for? Leak but not return seems surprising.

mwyrzykowski replied:
It's used in `WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer()` I think but would need to defer to Dan

mwyrzykowski replied:
I think its to ensure the textures get deleted when the framebuffer's destructor runs

hortont424 commented on Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm:
```diff
@@ -80,6 +121,77 @@ void destroyPbufferAndDetachIOSurface(EGLDisplay display, void* handle)
+
+    RetainPtr<MTLRasterizationRateLayerDescriptor> layerDescriptorLeft = [[MTLRasterizationRateLayerDescriptor alloc] initWithSampleCount:mtlScreenSize];
+
+    memcpy([layerDescriptorLeft horizontalSampleStorage], horizontalSamplesLeft.data(), horizontalSamplesLeft.size_bytes());
```
I definitely don't know enough to know that these are safe (and don't obviously see assertions that the sizes of these things are equal)

mwyrzykowski replied:
it is a few lines up:
```
    if (horizontalSamplesLeft.size() > mtlScreenSize.width || horizontalSamplesRight.size() > mtlScreenSize.width || verticalSamples.size() > mtlScreenSize.height)
        return nullptr;
```

mwyrzykowski replied:
oh I think I should also add: `|| !layerDescriptorLeft.get() || !layerDescriptorRight.get()`

in case the descriptor allocation fails

hortont424 reviewed (approved):
This is a half-review-half-rubber-stamp because I'm definitely not the best reviewer for this code, but predicated on the case that multiple other reviewers have... reviewed (and written)... in the past.

djg committed:
[WebXR] Refactor WebXROpaqueFramebuffer attachment logic
https://bugs.webkit.org/show_bug.cgi?id=269683
rdar://123201364

Reviewed by Tim Horton.

Patches are from Dan and I added some small bug fixes.

Refactor WebXROpaqueFramebuffer handling of framebuffer attachments to support
Shared or Layered layout.

There are a few combinations of context attributes that WebXROpaqueFramebuffer
handles:

if `antialias`:
   Create internal MSAA color buffer
   if `depth/stencil`
      Create internal MSAA depth/stencil buffer
if `!antialias`:
   Draw to externally provided color texture
   if `depth/stencil`
       if `!externally provided depth/stencil`
           Create internal depth/stencil buffer

Prior to this change, the code handled these combinations in `StartFrame(...)` &
`SetupFramebuffer(...)` with a loose collection of state.

This change simplifies the old assumption, dropping support for external
Textures, standardizing on Renderbuffers, and introduces helpers for keeping
related state together via `WebXRExternalRenderbuffer` and
`WebXRAttachmentSet`.

WebXRExternalRenderbuffer represents an externally provided Renderbuffer through
the OpenGL renderbuffer object and EGLImage pointer.

WebXRAttachmentSet represents a group of related color, depth and stencil
buffers. These can be "internal" or "external" via `WebXRAttachments` &
`WebXRExternalAttachments`.

Object naming has been changed to consistently use `draw` for the target of
WebGL rendering commands and `display` for the target presented to the user via
a WebXR compositor system. To simplify the handling of attachments, the output
of WebGL commands are always "blitted" from the draw target to display target,
removing the optimal path in the previous code. This path will not be taken on
any compositors supported on Cocoa and will be redressed in a future patch.

Due to limitations of `glBlitFramebuffer`, blitting to `Layered` layout from an
MSAA enabled attachment set requires the use of an intermediate, resolved path
represented by the `resolve` prefixed `m_resolvedAttachments` & `m_resolveFBO`.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::ensure):
(WebCore::createAndBindCompositorBuffer):
(WebCore::WebXROpaqueFramebuffer::WebXROpaqueFramebuffer):
(WebCore::WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer):
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
(WebCore::WebXROpaqueFramebuffer::resolveMSAAFramebuffer):
(WebCore::WebXROpaqueFramebuffer::blitShared):
(WebCore::WebXROpaqueFramebuffer::blitSharedToLayered):
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer):
(WebCore::WebXROpaqueFramebuffer::allocateRenderbufferStorage):
(WebCore::WebXROpaqueFramebuffer::allocateAttachments):
(WebCore::WebXROpaqueFramebuffer::bindAttachments):
(WebCore::WebXRExternalRenderbuffer::destroyImage):
(WebCore::WebXRExternalRenderbuffer::release):
(WebCore::WebXRExternalRenderbuffer::leakObject):
(WebCore::createAndBindCompositorTexture): Deleted.
(WebCore::WebXROpaqueFramebuffer::allocateColorStorage): Deleted.
(WebCore::WebXROpaqueFramebuffer::allocateDepthStencilStorage): Deleted.
(WebCore::WebXROpaqueFramebuffer::bindColorBuffer): Deleted.
(WebCore::WebXROpaqueFramebuffer::bindDepthStencilBuffer): Deleted.
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
(WebCore::WebXRAttachmentSet::release):
(WebCore::WebXRAttachmentSet::leakObject):
(WebCore::WebXROpaqueFramebuffer::framebuffer const):
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
(WebCore::destroyFunc):
* Source/WebCore/platform/xr/PlatformXR.h:
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/UIProcess/XR/ios/PlatformXRARKit.mm:
(WebKit::ARKitCoordinator::renderLoop):

Metal: Override internal format of images

Add support for overriding the internal format when binding Metal
textures to EGL images.

Fixed: angleproject:8552
Change-Id: I364a88a2e608e462c5216c92927679814ab91ca5

The following changes make blitting work.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::createAndBindCompositorBuffer):
(WebCore::WebXROpaqueFramebuffer::startFrame):
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::createAndBindEGLImage):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::createAndBindEGLImage):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::createAndBindEGLImage):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::createAndBindEGLImage):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/RemoteGraphicsContextGLProxyCocoa.mm:

Helper utilites for creating, enabled and disabling the use of rasterization
rate maps in GPUP. Infrastructure inplace to support caching of map for Shared
and Layered layouts, but only Shared is currently supported.

* Source/WebCore/PAL/pal/spi/cocoa/MetalSPI.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/GraphicsTypesGL.h:
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm:
(WebCore::newRasterizationRateMap):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::createFoveation):
(WebCore::GraphicsContextGLCocoa::enableFoveation):
(WebCore::GraphicsContextGLCocoa::disableFoveation):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(createFoveation):
(enableFoveation):
(disableFoveation):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::createFoveation):
(WebKit::RemoteGraphicsContextGLProxy::enableFoveation):
(WebKit::RemoteGraphicsContextGLProxy::disableFoveation):
* Tools/Scripts/generate-gpup-webgl:

Refactor setup logic to be passed once. WebXROpaqueFramebuffer::setupFramebuffer
should extract the necessary data to create framebuffers from the information
provided in PlatformXR::FrameData::LayerSetupData.

Changed WebXRWebGLLayer to query WebXROpaqueFramebuffer for the viewport, since
WebXRWebGLLayer has no knowledge of the structure of the layout, dimensions,
etc.

Canonical link: https://commits.webkit.org/276825@main
